File: BAEKJOON/1321 - 군인/군인.py
import sys
input = sys.stdin.readline

# 세그먼트 트리(SegmentTree) 클래스 정의
class SegmentTree:
    def __init__(self, n):  # 세그먼트 트리의 초기화 메서드
        self.size = 1  # 초기 사이즈를 1로 설정
        # 주어진 n보다 크거나 같은 가장 가까운 2의 제곱수 찾기
        while self.size < n:  
            self.size *= 2  # 사이즈를 2배씩 늘림
        # 트리 배열을 0으로 초기화. 트리의 크기는 2 * self.size
        self.tree = [0] * (2 * self.size)
    
    # 초기 배열을 세그먼트 트리에 빌드하는 함수
    def build(self, arr):
        # 입력 배열을 세그먼트 트리의 리프 노드에 삽입
        for i in range(len(arr)):
            self.tree[self.size + i] = arr[i]
        # 리프 노드 이후의 모든 부모 노드들을 자식 노드들의 합으로 초기화
        for i in range(self.size - 1, 0, -1):
            self.tree[i] = self.tree[2*i] + self.tree[2*i+1]
    
    # 특정 인덱스의 값을 변경하는 함수
    def update(self, index, value):
        # 인덱스를 트리의 리프 노드 위치로 이동
        index += self.size
        # 리프 노드의 값을 value만큼 더함
        self.tree[index] += value
        # 부모 노드로 거슬러 올라가면서 값을 갱신
        while index > 1:
            index //= 2  # 부모 노드로 이동
            # 부모 노드는 두 자식 노드의 합으로 설정
            self.tree[index] = self.tree[2*index] + self.tree[2*index+1]
    
    # 특정 군번이 속한 부대를 찾는 함수
    def query(self, soldier):
        # soldier가 전체 군인의 수를 초과할 경우 -1을 반환
        if soldier > self.tree[1]:
            return -1
        index = 1  # 루트 노드에서 탐색 시작
        # 리프 노드에 도달할 때까지 반복
        while index < self.size:
            # 왼쪽 자식 노드의 값이 soldier보다 크거나 같으면 왼쪽으로 이동
            if soldier <= self.tree[2*index]:
                index = 2*index
            # 그렇지 않으면 오른쪽으로 이동하고, soldier 값을 왼쪽 자식의 값만큼 감소
            else:
                soldier -= self.tree[2*index]
                index = 2*index + 1
        # index는 리프 노드의 위치이므로, 이를 부대 번호로 변환하여 반환
        return index - self.size + 1

# 입력: 부대의 개수 n
n = int(input())

# 입력: 각 부대의 군사 수
soldiers = list(map(int, input().split()))

# 세그먼트 트리 생성 및 초기화
seg_tree = SegmentTree(n)
seg_tree.build(soldiers)

# 입력: 명령의 개수 m
m = int(input())

# m개의 명령을 처리
for _ in range(m):
    # 각 명령을 입력받음
    query = list(map(int, input().split()))
    
    # "1 i a" 형태의 명령: i번 부대에 a명 증감
    if query[0] == 1:
        seg_tree.update(query[1]-1, query[2])  # 부대 번호를 0-based index로 변환하여 업데이트
    
    # "2 i" 형태의 명령: i번 군번이 속한 부대 찾기
    else:
        print(seg_tree.query(query[1]))  # i번 군번이 속한 부대를 출력

File: BAEKJOON/1321 - 군인/군인.cs
using System;

class SegmentTree {
    public int size;           // 세그먼트 트리의 크기
    public int[] tree;         // 세그먼트 트리를 저장할 배열

    // 세그먼트 트리 초기화
    public SegmentTree(int n) {
        size = 1;
        // 주어진 n보다 크거나 같은 가장 가까운 2의 제곱수로 설정
        while (size < n) size *= 2;
        // 트리 배열을 0으로 초기화
        tree = new int[2 * size];
    }

    // 주어진 배열을 세그먼트 트리에 빌드
    public void Build(int[] arr) {
        // 입력 배열을 리프 노드에 삽입
        for (int i = 0; i < arr.Length; i++) {
            tree[size + i] = arr[i];
        }
        // 부모 노드들을 초기화
        for (int i = size - 1; i > 0; i--) {
            tree[i] = tree[2 * i] + tree[2 * i + 1];
        }
    }

    // 특정 인덱스의 값을 value만큼 업데이트
    public void Update(int index, int value) {
        // 리프 노드의 위치로 이동
        index += size;
        tree[index] += value; // 값 변경
        // 부모 노드로 거슬러 올라가며 값 갱신
        while (index > 1) {
            index /= 2; // 부모 노드로 이동
            tree[index] = tree[2 * index] + tree[2 * index + 1]; // 자식 노드의 합으로 갱신
        }
    }

    // 특정 군번이 속한 부대를 찾는 함수
    public int Query(int soldier) {
        if (soldier > tree[1]) return -1; // 군번이 전체 군인 수를 초과하면 -1 반환
        int index = 1; // 루트 노드에서 시작
        // 리프 노드에 도달할 때까지 반복
        while (index < size) {
            if (soldier <= tree[2 * index]) {
                index = 2 * index; // 왼쪽 자식으로 이동
            } else {
                soldier -= tree[2 * index]; // 왼쪽 자식 노드의 군사 수를 soldier에서 뺌
                index = 2 * index + 1; // 오른쪽 자식으로 이동
            }
        }
        return index - size + 1; // 부대 번호 반환
    }
}

class Program {
    static void Main(string[] args) {
        // 부대의 개수 입력
        int n = int.Parse(Console.ReadLine());
        // 각 부대의 군사 수 입력
        int[] soldiers = Array.ConvertAll(Console.ReadLine().Split(' '), int.Parse);

        // 세그먼트 트리 생성 및 초기화
        SegmentTree segTree = new SegmentTree(n);
        segTree.Build(soldiers);

        // 명령의 개수 입력
        int m = int.Parse(Console.ReadLine());
        for (int i = 0; i < m; i++) {
            // 명령어 입력
            string[] query = Console.ReadLine().Split(' ');
            if (query[0] == "1") {
                // "1 i a" 명령 처리
                int index = int.Parse(query[1]) - 1; // 0-based 인덱스 변환
                int value = int.Parse(query[2]);
                segTree.Update(index, value); // 업데이트 수행
            } else if (query[0] == "2") {
                // "2 i" 명령 처리
                int soldier = int.Parse(query[1]);
                Console.WriteLine(segTree.Query(soldier)); // 군번이 속한 부대 번호 출력
            }
        }
    }
}